### PR TITLE
Change EnsureSuccessStatusCode to not dispose response content

### LINKED ISSUE #26684: Change EnsureSuccessStatusCode to not dispose response content
Currently the `HttpResponseMessage.EnsureSuccessStatusCode` method always disposes the content of the response. The reason is explained in [a comment](https://github.com/dotnet/corefx/blob/bcc2940faeb58cc92a66c61ca404bed54a3a3bb3/src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs#L152-L155):

```csharp
// Disposing the content should help users: If users call EnsureSuccessStatusCode(), an exception is
// thrown if the response status code is != 2xx. I.e. the behavior is similar to a failed request (e.g.
// connection failure). Users don't expect to dispose the content in this case: If an exception is
// thrown, the object is responsible fore cleaning up its state.
```

However, this isn't always helpful. It's frequent for web APIs to return details about the error in the response body. If I want to map the error to a specific domain exception, I might write something along these lines:

```csharp
private async Task<T> GetResponseAsync<T>(HttpRequestMessage request)
{
    using (var response = await _client.SendAsync(request))
    {
        try
        {
            response.EnsureSuccessStatusCode();
            return await response.Content.ReadAsAsync<T>();
        }
        catch (HttpRequestException ex)
        {
            var errorDetails = await response.Content.ReadAsAsync<ErrorDetails>();
            throw CreateDomainExceptionFromDetails(errorDetails, ex);
        }
    }
}

private Exception CreateDomainExceptionFromDetails(ErrorDetails errorDetails, Exception innerException)
{
    // Some logic based on the error details
    // ...
}
```

However, this doesn't work, because when I try to read the content in the `catch` block, the content is already disposed. So I need to read the content in the `try` block instead, before the call to `EnsureSuccessStatusCode`:

```csharp
private async Task<T> GetResponseAsync<T>(HttpRequestMessage request)
{
    using (var response = await _client.SendAsync(request))
    {
        string content = null;
        try
        {
            content = await response.Content.ReadAsStringAsync();
            response.EnsureSuccessStatusCode();
            return JsonConvert.DeserializeObject<T>(content);
        }
        catch (HttpRequestException ex) when (content != null)
        {
            var errorDetails = JsonConvert.DeserializeObject<ErrorDetails>(content);
            throw CreateDomainExceptionFromDetails(errorDetails, ex);
        }
    }
}
```

It's less intuitive and makes the code harder to read

An `EnsureSuccessStatusCode(bool disposeContent)` overload would be helpful in this case. It would let the user state that they're taking responsibility for disposing the content (in the exemple above, the content would be disposed when the response is disposed at the end of the `using` statement).

(I'm aware that I could just check the `IsSuccessStatusCode` property, but I want the "normal" `HttpRequestException` as the inner exception of my domain exception)

I'm willing to submit a pull request if the proposal is approved.

### PR DESCRIPTION
**Pull request by davidsh**
Historically, this method has always disposed the response content if it was
throwing an exception. The reason documented in the comment really doesn't
make much sense, especially to just dispose the content but not the actual
response object. The developer pattern we advise when getting an HttpResponseMessage
is to wrap it in a 'using' statement or use an explicit .Dispose() call. And
disposing the HttpResponseMessage would always dispose the inner .Content
anyways.

The use case of #26684 is that the response content should be available regardless
if this method throws an exception or not. Developers normally don't expect an object
to get disposed (partially in this case) as a side-affect of calling a method on the
object. So, we're changing the behavior to be more consistent generally with other
.NET objects.

Fixes #26684

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/HttpResponseMessageTest.cs [hunk 1]
```diff
@@ -120,13 +120,13 @@ public void EnsureSuccessStatusCode_VariousStatusCodes_ThrowIfNot2xx()
         }
 
         [Fact]
-        public void EnsureSuccessStatusCode_AddContentToMessage_ContentIsDisposed()
+        public void EnsureSuccessStatusCode_AddContentToMessage_ContentIsNotDisposed()
```
Comment (stephentoub) [inline src/System.Net.Http/tests/FunctionalTests/HttpResponseMessageTest.cs]: Does this need to be skipped now on netfx, since it doesn't have the change?

Comment (davidsh) [reply src/System.Net.Http/tests/FunctionalTests/HttpResponseMessageTest.cs]: Yes.  Forgot about that. I'll update the test.

**Comment by cremor** [comment]
@davidsh @stephentoub The [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpresponsemessage.ensuresuccessstatuscode?view=netcore-2.1#System_Net_Http_HttpResponseMessage_EnsureSuccessStatusCode) should be updated to reflect this change.

**Comment by karelz** [comment]
@cremor what would you document and where? Can you submit PR in docs repo? (hit edit button on the MSDN docs page)

**Comment by davidsh** [comment]
>Remarks
The EnsureSuccessStatusCode method throws an exception if the HTTP response was unsuccessful. If the Content is not null, this method will also call Dispose to free managed and unmanaged resources.

This PR was merged after the .NET Core 2.1 branch was finalized.  So, the documentation is still correct.

However, starting **after** .NET Core 2.1, this sentence is no longer true.  It should be removed at that time from the documentation.
*"If the Content is not null, this method will also call Dispose to free managed and unmanaged resources."*

**Comment by cremor** [comment]
@karelz Sorry for being so unspecific. @davidsh perfectly described what I meant.

**Comment by karlzf** [comment]
> However, starting **after** .NET Core 2.1, this sentence is no longer true. It should be removed at that time from the documentation:
> _"If the Content is not null, this method will also call Dispose to free managed and unmanaged resources."_

I did not see the changes applied to 2.2 release: https://github.com/dotnet/corefx/blob/release/2.2/src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs. Can anyone confirm if the comment by @davidsh is still valid?

**Comment by stephentoub** [comment]
It's a bit confusing, but coreclr and corefx have very few changes between 2.1 and 2.2, only critical fixes. Most changes that went into corefx after 2.1, including this one, will be in 3.0, which is the next major rev of the runtime and libraries. 2.2 was primarily about improvements to ASP.NET.

**Comment by karlzf** [comment]
Thanks for clarifications @stephentoub

**Comment by davidsh** [comment]
I opened a documentation PR https://github.com/dotnet/dotnet-api-docs/pull/1971 to update the docs for .NET Core 3.0 to reflect this new behavior.

**Pull request merged**